### PR TITLE
Fix several CCE issues in query expressions that has string and xml types as the expected type.

### LINKED ISSUE #39641: [Bug][QueryExpression]:Looping a stream with error completion to create a concatenated xml fails with a type cast error
### Description

```ballerina
import ballerina/io;

type Invoice record {|
    string Sony_Item_ID;
    string Item_Model_Number;
    string Serial_Number;
    int Quantity;
|};

public function main() returns error? {
    stream<Invoice, io:Error?> fileReadCsvAsStream = check io:fileReadCsvAsStream("./test.csv");

    xml x = check from var item in fileReadCsvAsStream
        select xml `<Invoice>
                            <Sony_Item_ID>${item.Sony_Item_ID}</Sony_Item_ID>
                            <Item_Model_Number>${item.Item_Model_Number}</Item_Model_Number>
                            <Serial_Number>${item.Serial_Number}</Serial_Number>
                            <Quantity>${item.Quantity}</Quantity>
                        </Invoice>`;
}

```

Fails with

```
error: {ballerina}TypeCastError {"message":"incompatible types: 'any[]' cannot be cast to 'error'"}
        at bbb:main(bbb.bal:36)
```

When type is inferred creates an `xml:Element[]`

```ballerina
import ballerina/io;

type Invoice record {|
    string Sony_Item_ID;
    string Item_Model_Number;
    string Serial_Number;
    int Quantity;
|};

public function main() returns error? {
    stream<Invoice, io:Error?> fileReadCsvAsStream = check io:fileReadCsvAsStream("./test.csv");

    var x = check from var item in fileReadCsvAsStream
        select xml `<Invoice>
                            <Sony_Item_ID>${item.Sony_Item_ID}</Sony_Item_ID>
                            <Item_Model_Number>${item.Item_Model_Number}</Item_Model_Number>
                            <Serial_Number>${item.Serial_Number}</Serial_Number>
                            <Quantity>${item.Quantity}</Quantity>
                        </Invoice>`;

   io:println(x);  
    // [`<Invoice>
    //     <Sony_Item_ID>P00000000000029401</Sony_Item_ID>
    //     <Item_Model_Number>SEL70200GM</Item_Model_Number>
    //     <Serial_Number>1817535</Serial_Number>
    //     <Quantity>1</Quantity>
    // </Invoice>`,`<Invoice>
    //     <Sony_Item_ID>P00000000000029504</Sony_Item_ID>
    //     <Item_Model_Number>SEL100400GM</Item_Model_Number>
    //     <Serial_Number>1816034</Serial_Number>
    //     <Quantity>1</Quantity>
    // </Invoice>`]
}
```

When used an array as the collection in the query expression, concatenated `xml` is created successfully

```ballerina
public function main() returns error? {
    Invoice[] arr = [{Sony_Item_ID: "P00000000000029401", Item_Model_Number: "SEL70200GM", Serial_Number: "1817535", Quantity: 1}, {Sony_Item_ID: "P00000000000029401", Item_Model_Number: "SEL70200GM", Serial_Number: "1817535", Quantity: 1}];

    xml x = from var item in arr
        select xml `<Invoice>
                            <Sony_Item_ID>${item.Sony_Item_ID}</Sony_Item_ID>
                            <Item_Model_Number>${item.Item_Model_Number}</Item_Model_Number>
                            <Serial_Number>${item.Serial_Number}</Serial_Number>
                            <Quantity>${item.Quantity}</Quantity>
                        </Invoice>`;

   io:println(x);  
    // <Invoice>
    //     <Sony_Item_ID>P00000000000029401</Sony_Item_ID>
    //     <Item_Model_Number>SEL70200GM</Item_Model_Number>
    //     <Serial_Number>1817535</Serial_Number>
    //     <Quantity>1</Quantity>
    // </Invoice>`,`<Invoice>
    //     <Sony_Item_ID>P00000000000029504</Sony_Item_ID>
    //     <Item_Model_Number>SEL100400GM</Item_Model_Number>
    //     <Serial_Number>1816034</Serial_Number>
    //     <Quantity>1</Quantity>
    // </Invoice>
}
```

### Steps to Reproduce

_No response_

### Affected Version(s)

2201.4.0

### OS, DB, other environment details and versions

_No response_

### Related area

-> Compilation

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### LINKED ISSUE #40029: [Bug]: [Bug][QueryExpression]:Looping a stream with error completion to create a concatenated string fails with a type cast error
### Description

```
import ballerina/io;

type Invoice record {|
    string Sony_Item_ID;
    string Item_Model_Number;
    string Serial_Number;
    int Quantity;
|};
public function main() returns error? {
    error? err = trap foo();
    io:println(err);
}

function foo() returns error? {
    stream<Invoice, io:Error?> fileReadCsvAsStream = check io:fileReadCsvAsStream("test.csv");
    string:Char y = "b";
    string x = check from var _ in fileReadCsvAsStream
        select y;

    io:println(x);
}
```

Gives

```
error("{ballerina}TypeCastError",message="incompatible types: '(lang.query:Type & readonly)[] & readonly' cannot be cast to 'error'")
```

This is the `string` related issue that mention in #39641 (xml related issue).

### Steps to Reproduce

_No response_

### Affected Version(s)

_No response_

### OS, DB, other environment details and versions

_No response_

### Related area

-> Runtime

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### LINKED ISSUE #40030: [Bug]: Query expressions that have `string` or `xml` types as target types, evaluated a wrong resolved type in queryTypeChecker
### Description

Consider following codes,

1. XML as a target type
```
import ballerina/io;

type Invoice record {|
    string Sony_Item_ID;
    string Item_Model_Number;
    string Serial_Number;
    int Quantity;
|};

public function main() returns error? {
    error? err = trap foo();
    io:println(err);
}

function foo() returns error? {
    stream<Invoice, io:Error?> fileReadCsvAsStream = check io:fileReadCsvAsStream("test.csv");

        xml x = check from var item in fileReadCsvAsStream
            select xml `<Invoice>
                            </Invoice>`;
        io:println(x);
}
```

In here the, the `expType` of the query expression is evaluated as `xml:element`, But this should be `xml` type.

2. String as a target type

```
import ballerina/io;

type Invoice record {|
    string Sony_Item_ID;
    string Item_Model_Number;
    string Serial_Number;
    int Quantity;
|};
public function main() returns error? {
    error? err = trap foo();
    io:println(err);
}

function foo() returns error? {
    stream<Invoice, io:Error?> fileReadCsvAsStream = check io:fileReadCsvAsStream("test.csv");
    string:Char y = "b";
    string x = check from var _ in fileReadCsvAsStream
        select y;

    io:println(x);
}
```
In here the, the `expType` of the query expression is evaluated as `string:Char`, But this should be `string` type.

### Steps to Reproduce

_No response_

### Affected Version(s)

_No response_

### OS, DB, other environment details and versions

_No response_

### Related area

-> Compilation

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### LINKED ISSUE #40038: [Bug]: CCE in query with string as expected type
### Description

```
function queryWithString2() returns error? {
    stream<Book, error?> bookStream = [{ author: "Author 1", title: "Title 1" }, 
                                        {author: "Author 2", title: "Title 2" }].toStream();

    string strValue = check from Book _ in bookStream
        select <string:Char> "a";
}
public function main() {

    error? testQueryWithXmlResult2 = trap queryWithString2();
    if (testQueryWithXmlResult2 is error) {
        io:println(testQueryWithXmlResult2);
    }
}
```

Gives

```
Compiling source
        a.bal
WARNING [a.bal:(12:5,13:34)] unused variable 'strValue'

Running executable

[2023-03-30 10:45:00,661] SEVERE {b7a.log.crash} - class java.lang.String cannot be cast to class io.ballerina.runtime.api.values.BValue (java.lang.String is in module java.base of loader 'bootstrap'; io.ballerina.runtime.api.values.BValue is in unnamed module of loader 'app') 
java.lang.ClassCastException: class java.lang.String cannot be cast to class io.ballerina.runtime.api.values.BValue (java.lang.String is in module java.base of loader 'bootstrap'; io.ballerina.runtime.api.values.BValue is in unnamed module of loader 'app')
        at io.ballerina.runtime.internal.TypeChecker.getType(TypeChecker.java:367)
        at io.ballerina.runtime.internal.ErrorUtils.createNumericConversionError(ErrorUtils.java:148)
        at io.ballerina.runtime.internal.TypeConverter.stringToChar(TypeConverter.java:930)
        at io.ballerina.runtime.internal.TypeConverter.anyToChar(TypeConverter.java:992)
        at a.queryWithString2(a.bal:12)
        at a.main(a.bal:17)
        at $_init.$moduleExecute(.)
        at $_init.$lambda$$moduleExecute$(.)
        at io.ballerina.runtime.internal.scheduling.SchedulerItem.execute(SchedulerItem.java:54)
        at io.ballerina.runtime.internal.scheduling.Scheduler.run(Scheduler.java:306)
        at io.ballerina.runtime.internal.scheduling.Scheduler.runSafely(Scheduler.java:273)
        at java.base/java.lang.Thread.run(Thread.java:829)
```

### Steps to Reproduce

_No response_

### Affected Version(s)

_No response_

### OS, DB, other environment details and versions

_No response_

### Related area

-> Compilation

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by SasinduDilshara**
Fixes #39641 
Fixes #40029 
Fixes #40030 
Fixes #40038 

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/query/StringQueryExpressionTest.java [hunk 1]
```diff
@@ -48,6 +48,11 @@ public void testSimpleQueryExprForStringResult() {
         Assert.assertEquals(returnValues.toString(), "Alex Ranjan John ");
     }
 
+    @Test(description = "Test simple query expression with string result")
```
Comment (gimantha) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/query/StringQueryExpressionTest.java:51]: Lets add a negative test for ambiguous types also.

 ```bal
  string:Char[]|string strValue = check from Book _ in bookStream
        select chr;
```

Comment (gimantha) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/query/StringQueryExpressionTest.java:51]: We can add it to `xml|xml[]` case too

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/QueryDesugar.java [hunk 2]
```diff
@@ -297,13 +297,13 @@ BLangStatementExpression desugar(BLangQueryExpr queryExpr, SymbolEnv env,
             onConflictExpr = null;
         } else {
             BType refType = Types.getReferredType(queryExpr.getBType());
-            if (isXml(refType)) {
+            if (isXml(types.getSafeType(refType, true, true))) {
```
Comment (pcnfernando) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/QueryDesugar.java:300]: Get the safe type commonly for both cases

File: tests/jballerina-unit-test/src/test/resources/test-src/query/string-query-expression.bal [hunk 3]
```diff
@@ -41,6 +41,64 @@ function testSimpleQueryExprForStringResult() returns string {
...
+
+    e = trap simpleQueryExprForStringResult2();
+    assertFalse(e is error);
+}
+
+function simpleQueryExprForStringResult() returns error? {
+    string:Char chr = "a";
+    BookGenerator bookGenerator = new ();
+    stream<Book, error?> bookStream = new (bookGenerator);
+
```
Comment (sanjana) [inline tests/jballerina-unit-test/src/test/resources/test-src/query/string-query-expression.bal:79]:
```suggestion
```

File: tests/jballerina-unit-test/src/test/resources/test-src/query/string-query-expression.bal [hunk 4]
```diff
@@ -41,6 +41,64 @@ function testSimpleQueryExprForStringResult() returns string {
...
+    string strValue = check from Book _ in bookStream
+        select chr;
+
+    string expectedValue = "aa";
+
```
Comment (sanjana) [inline tests/jballerina-unit-test/src/test/resources/test-src/query/string-query-expression.bal:84]:
```suggestion
```

File: tests/jballerina-unit-test/src/test/resources/test-src/query/string-query-expression.bal [hunk 5]
```diff
@@ -41,6 +41,64 @@ function testSimpleQueryExprForStringResult() returns string {
...
+function simpleQueryExprForStringResult2() returns error? {
+    stream<Book, error?> bookStream = [{ author: "Author 1", title: "Title 1" },
+                                        {author: "Author 2", title: "Title 2" }].toStream();
+
+    string strValue = check from Book _ in bookStream
+        select <string:Char> "a";
+
+    string expectedValue = "aa";
+
```
Comment (sanjana) [inline tests/jballerina-unit-test/src/test/resources/test-src/query/string-query-expression.bal:97]:
```suggestion
```

File: tests/jballerina-unit-test/src/test/resources/test-src/query/xml-query-expression.bal [hunk 6]
```diff
@@ -87,6 +87,185 @@ function testSimpleQueryExprForXML3() returns xml {
...
+
+function simpleQueryExprForXML4() returns error? {
+    BookGenerator bookGenerator = new ();
+    stream<Book, error?> bookStream = new (bookGenerator);
+
+    xml xmlValue = check from Book book in bookStream
+        select xml `${book.title} - ${book.author}.`;
+
+    xml expectedValue = xml `Title 1 - Author 1.Title 2 - Author 2.`;
+
```
Comment (sanjana) [inline tests/jballerina-unit-test/src/test/resources/test-src/query/xml-query-expression.bal:213]:
```suggestion
```

File: tests/jballerina-unit-test/src/test/resources/test-src/query/xml-query-expression.bal [hunk 7]
```diff
@@ -87,6 +87,185 @@ function testSimpleQueryExprForXML3() returns xml {
...
+function simpleQueryExprForXML5() returns error? {
+    BookGenerator bookGenerator = new ();
+    stream<Book, error?> bookStream = new (bookGenerator);
+
+    xml xmlValue = check from Book book in bookStream
+        select xml `<!--${book.title} - ${book.author}-->`;
+
+    xml expectedValue = xml `<!--Title 1 - Author 1--><!--Title 2 - Author 2-->`;
+
```
Comment (sanjana) [inline tests/jballerina-unit-test/src/test/resources/test-src/query/xml-query-expression.bal:227]:
```suggestion
```

File: tests/jballerina-unit-test/src/test/resources/test-src/query/xml-query-expression.bal [hunk 8]
```diff
@@ -87,6 +87,185 @@ function testSimpleQueryExprForXML3() returns xml {
...
+function simpleQueryExprForXML6() returns error? {
+    BookGenerator bookGenerator = new ();
+    stream<Book, error?> bookStream = new (bookGenerator);
+
+    xml xmlValue = check from Book _ in bookStream
+        select xml `<?target data?>`;
+
+    xml expectedValue = xml `<?target data?><?target data?>`;
+
```
Comment (sanjana) [inline tests/jballerina-unit-test/src/test/resources/test-src/query/xml-query-expression.bal:241]:
```suggestion
```

File: tests/jballerina-unit-test/src/test/resources/test-src/query/xml-query-expression.bal [hunk 9]
```diff
@@ -87,6 +87,185 @@ function testSimpleQueryExprForXML3() returns xml {
...
+    assertTrue(xmlValue is xml);
+    assertTrue(xmlValue is xml<xml:Element>);
+    assertEquality(xmlValue, expectedValue);
+}
+
+function simpleQueryExprForXML3() returns error? {
+    BookGenerator bookGenerator = new ();
+    stream<Book, error?> bookStream = new (bookGenerator);
+
```
Comment (sanjana) [inline tests/jballerina-unit-test/src/test/resources/test-src/query/xml-query-expression.bal:185]:
```suggestion
```

**Review by sanjana — CHANGES_REQUESTED** [review]
Minor formatting suggestions.
Look for all places where applicable.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/QueryTypeChecker.java [hunk 10]
```diff
@@ -323,12 +323,19 @@ void solveSelectTypeAndResolveType(BLangQueryExpr queryExpr, BLangExpression sel
...
             case TypeTags.XML_COMMENT:
             case TypeTags.XML_ELEMENT:
             case TypeTags.XML_PI:
             case TypeTags.XML_TEXT:
                 selectType = checkExpr(selectExp, env, type, data);
+                if (types.isAssignable(selectType, symTable.xmlType)) {
+                    resolvedType = getResolvedType(new BXMLType(selectType, null), type, isReadonly, env);
+                    break;
+                }
                 resolvedType = selectType;
```
Comment (KavinduZoysa) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/QueryTypeChecker.java:339]: Let's use if-else here.

File: tests/jballerina-unit-test/src/test/resources/test-src/query/string-query-expression.bal (hunk 3 above)
Comment (SasinduDilshara) [reply tests/jballerina-unit-test/src/test/resources/test-src/query/string-query-expression.bal:79]: These are intentionally put to improve the readability of the code.

```
test function () {
  < var declaration>
  empty line
  <query declaration>
  empty line
  <assertions>
}
```

Wdyt?

File: tests/jballerina-unit-test/src/test/resources/test-src/query/xml-query-expression.bal [hunk 11]
```diff
@@ -87,6 +87,185 @@ function testSimpleQueryExprForXML3() returns xml {
...
+    e = trap simpleQueryExprForXML7();
+    assertFalse(e is error);
+}
+
+function simpleQueryExprForXML() returns error? {
+    stream<Book, error?> bookStream = [{ author: "Author 1", title: "Title 1" },
+                                        {author: "Author 2", title: "Title 2" }].toStream();
+
+    xml xmlValue = check from Book book in bookStream
```
[the remark this answers is not in the document]
Comment (SasinduDilshara) [reply tests/jballerina-unit-test/src/test/resources/test-src/query/xml-query-expression.bal:150]: added.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/QueryDesugar.java (hunk 2 above)
Comment (SasinduDilshara) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/QueryDesugar.java]: Added. 35ec75614ab6aa67688f321ffeb8b3c19441b285

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/query/StringQueryExpressionTest.java (hunk 1 above)
Comment (SasinduDilshara) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/query/StringQueryExpressionTest.java:52]: Added tests for ambiguous situations in both `string`,`xml`.
But those tests will uncomment after fixing #40012

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/QueryTypeChecker.java (hunk 10 above)
Comment (SasinduDilshara) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/QueryTypeChecker.java]: Updated. 35ec75614ab6aa67688f321ffeb8b3c19441b285

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/query/StringQueryExpressionTest.java [hunk 12]
```diff
@@ -172,6 +178,15 @@ public void testQueryExprWithLimitForStringResultV2() {
...
+    // public void testNegativeQueryExprForXML() {
+    //     int index = 0;
+    //     validateError(negativeResult, index++, "ambiguous type '[string:Char, string:Char]'", 46, 16);
+    //     validateError(negativeResult, index++, "ambiguous type '[string:Char, string:Char, string:Char]'", 48, 16);
+    //     Assert.assertEquals(negativeResult.getErrorCount(), index);
+    // }
+
     @AfterClass
     public void tearDown() {
         result = null;
```
Comment (sanjana) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/query/StringQueryExpressionTest.java:192]: Need to set the `negativeResult` to `null` as well.
```suggestion
        result = null;
        negativeResult = null;
```

File: tests/jballerina-unit-test/src/test/resources/test-src/query/string-query-expression-negative.bal [hunk 13]
```diff
@@ -0,0 +1,49 @@
+// Copyright (c) 2023 WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+//
+// WSO2 Inc. licenses this file to you under the Apache License,
```
Comment (sanjana) [inline tests/jballerina-unit-test/src/test/resources/test-src/query/string-query-expression-negative.bal:3]: Please check the licenses header.

```suggestion
// Copyright (c) 2023 WSO2 LLC. (http://www.wso2.org) All Rights Reserved.
//
// WSO2 LLC. licenses this file to you under the Apache License,
```

File: tests/jballerina-unit-test/src/test/resources/test-src/query/string-query-expression.bal [hunk 14]
```diff
@@ -336,3 +400,25 @@ function testQueryExprWithLimitForStringResultV2() returns string {
...
 }
+
+function assertEquality(any|error actual, any|error expected) {
+    if expected is anydata && actual is anydata && expected == actual {
+        return;
+    }
+
+    if expected === actual {
+        return;
+    }
```
Comment (sanjana) [inline tests/jballerina-unit-test/src/test/resources/test-src/query/string-query-expression.bal:411]: Better if the brackets are added here.

```suggestion
    if (expected is anydata && actual is anydata && expected == actual) {
        return;
    }

    if (expected === actual) {
        return;
    }
```

File: tests/jballerina-unit-test/src/test/resources/test-src/query/xml-query-expression.bal [hunk 15]
```diff
@@ -87,6 +87,262 @@ function testSimpleQueryExprForXML3() returns xml {
...
+|};
+
+class BookGenerator {
+    int i = 0;
+
+    public isolated function next() returns record {|Book value;|}|error? {
+        self.i += 1;
+        if (self.i < 0) {
+            return error("Error");
+        } else if (self.i >= 3) {
```
Comment (sanjana) [inline tests/jballerina-unit-test/src/test/resources/test-src/query/xml-query-expression.bal:102]: Can we get rid off the `else` here?

```suggestion
        } 
        
        if (self.i >= 3) {
```

Comment (SasinduDilshara) [reply tests/jballerina-unit-test/src/test/resources/test-src/query/xml-query-expression.bal:102]: Imo, it is better to add else if

File: tests/jballerina-unit-test/src/test/resources/test-src/query/string-query-expression.bal (hunk 14 above)
Comment (SasinduDilshara) [reply tests/jballerina-unit-test/src/test/resources/test-src/query/string-query-expression.bal:411]: I guess the best practice is to not add brackets in ballerina. @pcnfernando @LakshanWeerasinghe wdyt?

**Comment by pcnfernando** [comment]
@SasinduDilshara please check on the buld failure

**Comment by pcnfernando** [comment]
@sanjana please enable the tests that will be addressed in https://github.com/ballerina-platform/ballerina-lang/issues/40012 once this PR is merged

**Review by pcnfernando — APPROVED** [review]

**Comment by pcnfernando** [comment]
@SasinduDilshara can you please check and port the fix to applicable patch branches

**Pull request merged**